test/langtools/tools/javap/InnerClass.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2014, 2017, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8032869
+ * @summary javap should render inner classes in a friendly way
+ * @modules jdk.jdeps/com.sun.tools.javap
+ */
+
+import java.io.*;
+import java.util.function.Function;
+import javax.lang.model.element.ElementKind;
+
+public class InnerClass {
+    public static void main(String... args) throws Exception {
+        new InnerClass().run();
+    }
+
+    void run() throws Exception {
+        String testClasses = System.getProperty("test.classes");
+        String out = javap("-v", "-classpath", testClasses, A.class.getName());
+
+        String nl = System.getProperty("line.separator");
+        out = out.replaceAll(nl, "\n");
+
+        if (out.contains("\n\n\n"))
+            error("double blank line found");
+
+        expect(out,
+                "InnerClasses:\n" +
+                        "  public static #17= #7 of #14;           // A=class InnerClass$A of class InnerClass\n" +
```

Comment:
Hard-coded expected output (including exact CP indexes) does not validate actual output against actual javap, but rather conserves what we produce now (no matter if correct or wrong).

Reply:
I copied this concept from other javap tests. Would you say that a regex match would be a better option?

Reply:
Yes, regexp based comparison is one of the options.
I would prefer to  dedicate a new package for these tests, with stored original "golden" file outputs from the actual JDK javap as a reference, and tests performing comparison. As the CP indexes (and order) would not match, it would require to mask them, or perform any other form of normalized comparison (with focus on compatible syntax).
Similar (however a bit more complex) normalized comparison of the whole classes is for example implemented in [helpers/ClassRecord.java](https://github.com/openjdk/jdk-sandbox/blob/classfile-api-branch/test/jdk/jdk/classfile/helpers/ClassRecord.java).

These compatibility test cases are useful only for transition period and they become obsolete when integrated into JDK.

Reply:
I followed the same pattern as all other javap tests now where only fragments are checked for their presence. I would argue that creating such a test framework is too big of a task for a PR like this. I registered these added tests on the classfile API ticket, if that's an option?

I agree with you that some form of better test framework for *javap* would be desirable. Test coverage is rather random and only checks for fragments are currently supported. For comparing the output of two javap versions, I do however not think an automated test is necessary if these tests are supposed to be removed again. I currently use two *javap* on the command line and pipe the output to the *diff* tool. I ran this over a range of classes of my test repository where I attempted to include all class file elements that exist and that's how I found the issues I am fixing in this PR. I hope there is nothing else missing for now.

My updated PR: https://github.com/openjdk/jdk-sandbox/pull/25